src/backend/distributed/executor/executor.c

```diff
@@ -538,8 +538,15 @@ FinishDistributedExecution(DistributedExecution *execution)
 			/* TODO: should we keep this in the state machine? or cancel? */
 			ClearResults(connection, false);
 		}
-
-		/* TODO: can we be in SENT_COMMAND? */
+		else if (transactionState == REMOTE_TRANS_SENT_COMMAND)
+		{
+			/*
+			 * Result of a command contains errors. So we change the state
+			 * to clear results and reset state, so we can process future
+			 * ROLLBACK TO SAVEPOINTs.
+			 */
+			transaction->transactionState = REMOTE_TRANS_CLEARING_RESULTS;
```

Comment:
As discussed, I think the main effect of this is that it puts the transaction back into a state from which it is safe to re-enter the unified executor, whereas `REMOTE_TRANS_SENT_COMMAND` is not safe. In case of a rollback to savepoint we already clear the results in `FinishRemoteTransactionSavepointRollback` so we do not actually need to re-enter that state. However, `FinishRemoteTransactionSavepointRollback` currently neglects to update `transactionState` to `REMOTE_TRANS_STARTED`.

Reply:
Removing this and changing the transaction state in `FinishRemoteTransactionSavepointRollback`.